OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

thomasmueller opened the pull request.

reschke committed:
OAK-8769: oak-auth-ldap pom needs maintenance - backed out due to build errors

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876648 13f79535-47bb-0310-9956-ffa450edef68

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneDocumentMaker.java:
```diff
@@ -54,6 +55,10 @@
 
+    private static final String DYNAMIC_BOOST_TAG_NAME = "name";
+    private static final String DYNAMIC_BOOST_TAG_CONFIDENCE = "confidence";
+    private static final String DYNAMIC_BOOST_SPLIT_REGEX = "[:/]";
```
Would It make sense to get these 3 properties DYNAMIC_BOOST_TAG_NAME, DYNAMIC_BOOST_TAG_CONFIDENCE and DYNAMIC_BOOST_SPLIT_REGEX  from the index definition ? (This would make this more generic in case these nodes are generated by some service other than the dynamic tag one - that service could have a different property name for a value that can be used as boost.)

thomasmueller replied:
Sure it is possible. However, the algorithm is so specific anyway, I don't think we can reuse it ever. So I see no point in making it configurable: it would just add complexity, without solving anything. So I'm against that :-)

nit0906 commented on oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/spi/editor/FulltextDocumentMaker.java:
```diff
@@ -101,6 +101,8 @@ public FulltextDocumentMaker(@Nullable FulltextBinaryTextExtractor textExtractor
 
     protected abstract void indexTypedProperty(D doc, PropertyState property, String pname, PropertyDefinition pd, int index);
 
+    protected abstract boolean indexDynamicBoost(D doc, PropertyDefinition pd, NodeState nodeState, String propertyName);
```
could we maybe change the method name to something else ? Since for elastic, we would actually not be doing index time boosting, and just adding a value that can later on be used for query time boosting.
Maybe supportDynamicBoost or manageDynamicBoost and then add a comment what the method expects from the implementations that extend this class
(Like either do an index time boost or add some property for the boost value which can be used during query time boost )

thomasmueller replied:
Well all other methods are called "index" so this prefix I would like to keep. "To index" means to add data to the index.

The term "dynamic boost": I called the "feature" that way. I think the term should reflect what it is doing on a high level, and not how it's implemented. An alternative: we could use "dynamic confidence", if "boost" is too much a Lucene / Elasticsearch term. But I don't think "confidence" is better than "boost"... It is about boosting some documents. Whether that's done at indexing time or query time is an implementation detail. I think the Asset team should decide on how to call the feature.

So "indexDynamicBoost" means: Store the data in the index that is needed for the "dynamic boost" feature.

nit0906 commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/IndexFieldProviderImpl.java:
```diff
@@ -0,0 +1,112 @@
+/**
+ * An example index field provider.
+ */
+public class IndexFieldProviderImpl implements IndexFieldProvider {
```
Should we mark the interface IndexFieldProvider as deprecated if we plan on removing this altogether at some point of time in future ?

thomasmueller replied:
Yes, we should mark it as deprecated (we should have done that before).

thomasmueller replied:
Now marked both IndexFieldProvider and FulltextQueryTermsProvider as deprecated.

sdmcraft replied:
Ah I see. I should have seen the full file paths. This is indeed in `test`.

nit0906 commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/DynamicBoostTest.java:
```diff
@@ -0,0 +1,236 @@
+        
+        Tree t = createTestNodes(root);
+
+        AsyncLogFilter filter = new AsyncLogFilter(LuceneDocumentMaker.class);
```
There is a LogCustomizer in oak-commons to do log related verifications. Maybe we can use that here. That also makes use of ILoggingEvent.

thomasmueller replied:
Oh I didn't know about LogCustomizer, I'll check.

thomasmueller replied:
Used LogCustomizer now. This simplified the test case quite a bit!

nit0906 reviewed (commented):
Added a few comments.

fabriziofortino commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/FulltextQueryTermsProviderImpl.java:
```diff
@@ -0,0 +1,150 @@
+        query.add(new TermQuery(term), BooleanClause.Occur.SHOULD);
+        LOG.debug("Added query term: {}", text.toLowerCase());
+
+        //De-boosting smart tags based query. CQ-4194419
+        query.setBoost(0.0001f);
+        return query;
```
I understand why we need this. Wondering if we can be smarter here. Moreover, should we move the deboost value in the index def?

thomasmueller replied:
The class isn't actually used currently, I could remove it from the patch. It will be needed once we implement the "query" part of this feature in Oak.

But I wouldn't currently want to change how it's implemented... It is what is in use _currently_. We want to replace it with something else, for sure; but we should keep it in the test source folder so we can compare the new and the old implementation. We can change the behavior slightly in the Oak feature, but not here.

fabriziofortino commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/FulltextQueryTermsProviderImpl.java:
```diff
@@ -0,0 +1,150 @@
+                    "maximum permissible value of {}", charTerms.size(), MAX_QUERY_SIZE);
+            return null;
+        }
+        List<String> fragments = prepareFragments(charTerms);
```
it would be nice to add some documentation here on why we need to create all these fragments. An example of input `query` with the resulting terms would be great. The logic is quite articulated. I am wondering if we can simplify this with some more sophisticated query (maybe it is not possible in lucene but it could be in elastic)

fabriziofortino commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/FulltextQueryTermsProviderImpl.java:
```diff
@@ -0,0 +1,150 @@
+        query.add(new TermQuery(term), BooleanClause.Occur.SHOULD);
+        LOG.debug("Added query term: {}", text.toLowerCase());
+
+        //De-boosting smart tags based query. CQ-4194419
+        query.setBoost(0.0001f);
+        return query;
+
```
I understand why we need this. Wondering if we can be smarter here. Moreover, should we move the deboost value in the index def?

fabriziofortino reviewed (commented):
LGTM. I just added a couple of questions

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

nit0906 reviewed (approved):
lgtm

reschke committed:
OAK-9010: Standardize handling of system properties

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876746 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9016: link slf4j javadocs

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876747 13f79535-47bb-0310-9956-ffa450edef68

sdmcraft commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneDocumentMaker.java:
```diff
@@ -331,4 +336,69 @@ protected void indexSimilarityBinaries(Document doc, PropertyDefinition pd, Blob
     }
+    
+    @Override
+    protected boolean indexDynamicBoost(Document doc, PropertyDefinition pd, NodeState nodeState, String propertyName) {
```
This method seems to assume a specific structure such as that of a node having a bunch of children with properties DYNAMIC_BOOST_TAG_NAME and DYNAMIC_BOOST_TAG_CONFIDENCE. How is it ensured that this method gets called for only where it ought to be called?

thomasmueller replied:
> a specific structure such as that of a node having a bunch of children with properties DYNAMIC_BOOST_TAG_NAME and DYNAMIC_BOOST_TAG_CONFIDENCE

Yes, exactly.

> How is it ensured that this method gets called for only where it ought to be called?

This is ensured by the index definition. To enable the feature, add a property to be indexed, e.g.:

    /oak:index/damAssetLucene-...
        + indexRules
            + dam:Asset
                + properties
                    + predictedTags
                        - dynamicBoost = true (Boolean)
                        - propertyIndex = true (Boolean)
                        - isRegexp = true (Boolean)
                        - name = jcr:content/metadata/predictedTags/.* (String)

The "dynamicBoost = true" enables this feature. The method is then called for all dam:Asset nodes that have a property jcr:content/metadata/predictedTags/.* (any property). That includes jcr:content/metadata/predictedTags/jcr:primaryType, meaning the code is called when the node predictedTags is added. Unlike how it was done with the IndexFieldProvider, the code is also called if any of property in that node is modified. I don't think that's needed right now however.



dynamicBoost = true (Boolean)
propertyIndex = true
name = 
isRegexp = true (Boolean)

sdmcraft replied:
Ok. The `dynamicBoost` and `propertyIndex` look like new additions. Below is from the current index definition of `/oak:index/damAssetLucene/indexRules/dam:Asset/properties/predictedTags` 
![Screenshot 2020-04-20 at 6 10 24 PM](https://user-images.githubusercontent.com/1191451/79752770-73a4b900-8332-11ea-9cbb-4bb36554ca96.png)

reschke committed:
OAK-9017: use SystemPropertySupplier in RDBConnectionHandler

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876748 13f79535-47bb-0310-9956-ffa450edef68

sdmcraft commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/FulltextQueryTermsProviderImpl.java:
```diff
@@ -0,0 +1,150 @@
+/**
+ * An example fulltext query terms provider.
+ */
+public class FulltextQueryTermsProviderImpl implements FulltextQueryTermsProvider {
```
Is there a specific requirement to have this  implementation in here ? Or can it reside with the application code?

thomasmueller replied:
Right now it can reside in the application code. I added this class here for later, when we have time to also move that part to Oak. Moving the code to Oak is not needed for the Lucene case, but it will be needed for Elasticsearch.

sdmcraft replied:
ok

reschke committed:
OAK-9018: use SystemPropertySupplier in RDBVersionGCSupport

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876749 13f79535-47bb-0310-9956-ffa450edef68

sdmcraft commented on oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/IndexFieldProviderImpl.java:
```diff
@@ -0,0 +1,112 @@
+    }
+    
+    @Override
+    public @NotNull Iterable<Field> getAugmentedFields(String path, NodeState nodeState, NodeState indexDefinition) {
```
`LuceneDocumentMaker#indexDynamicBoost` seems to doing the same logic.  Why do we need it at 2 places?

thomasmueller replied:
This is the test case. In the test case, on purpose I also added a test that uses the old logic (using the IndexFieldProvider), because so far there was no such test case. At no test case with the logic that is now implemented in Oak.

sdmcraft replied:
ok

reschke committed:
OAK-9019: use SystemPropertySupplier in RDBMissingLastRevSeeker

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876750 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9020: use SystemPropertySupplier in RDBDocumentStoreDB

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876751 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9021: use SystemPropertySupplier in RDBDataSourceFactory

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876752 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9022: use SystemPropertySupplier in ClusterNodeInfo

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876753 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9023: use SystemPropertySupplier in RDBDocumentStore

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876754 13f79535-47bb-0310-9956-ffa450edef68

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider - deprecate IndexFieldProvider and FulltextQueryTermsProvider

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider - deprecate IndexFieldProvider and FulltextQueryTermsProvider

sdmcraft commented:
@thomasmueller , is there a way I can try the changes on an instance? Maybe if you can provide build steps, I can try deploying it locally.

reschke committed:
OAK-9010: Standardize handling of system properties (add support for Long)

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876760 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9025: use SystemPropertySupplier in DocumentNodeStoreService

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876762 13f79535-47bb-0310-9956-ffa450edef68

Commit:
OAK-8769: oak-auth-ldap pom needs maintenance

Updated to org.apache.directory.api.api-all-2.0.1.
The LDAP test server now uses a separate ClassLoader because it still uses v1.


git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876763 13f79535-47bb-0310-9956-ffa450edef68

Commit:
OAK-9008 | GET 1587444481700._test1-0/_search (Patch by Fabrizio)

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876774 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-8769: oak-auth-ldap pom needs maintenance - fix svn:eol:style

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876776 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-9028: upgrade to mockito-core 3.3.3

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876785 13f79535-47bb-0310-9956-ffa450edef68

Commit:
OAK-9015 | ElasticserachIndexWriter#close method implementation should follow the underlying contract's doc

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876791 13f79535-47bb-0310-9956-ffa450edef68

reschke committed:
OAK-8995: remove package from Export-Package

git-svn-id: https://svn.apache.org/repos/asf/jackrabbit/oak/trunk@1876800 13f79535-47bb-0310-9956-ffa450edef68

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/spi/FulltextQueryTermsProvider.java:
```diff
@@ -24,8 +24,14 @@
+ *             as soon as an alternative is available. See also
+ *             IndexFieldProvider.
  */
 public interface FulltextQueryTermsProvider {
```
We should add @Deprecated annotation to the interface as well along with the @deprecated tag in the comment (I think we don't get a deprecation warning during compile time if we don't use the annotation as below)- 

@Deprecated
public interface FulltextQueryTermsProvider {

nit0906 commented on oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/spi/IndexFieldProvider.java:
```diff
@@ -25,6 +25,9 @@
+ * @deprecated This interface exposes Lucene API directly - it will be removed soon.
+ * Use the "dynamicBoost" feature instead.
  */
 public interface IndexFieldProvider {
```
We should add @Deprecated annotation to the interface as well along with the @deprecated tag in the comment (I think we don't get a deprecation warning during compile time if we don't use the annotation as below)- 

@Deprecated
public interface FulltextQueryTermsProvider {

nit0906 reviewed (changes requested):
deprecated annotation needed in the interfaces along with the deprecation tag in comments.

thomasmueller commented:
@nit0906  you are right of course! I will add that, test it, and then commit the change (actually the annotation is uppercase, `@Deprecated`)

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

thomasmueller committed:
OAK-8971 Indexing: dynamic boost, as an alternative to IndexFieldProvider

thomasmueller committed:
Merge branch 'OAK-8971' of https://github.com/oak-indexing/jackrabbit-oak into OAK-8971

nit0906 reviewed (approved):
👍
